sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/xds/XdsRouterHandler.java

```diff
@@ -0,0 +1,231 @@
+            Set<ServiceInstance> serviceInstances = getServiceInstanceOfCluster(clusterLoadAssigment);
+            return serviceInstances.isEmpty() ? serviceDiscovery.getServiceInstance(serviceName) : serviceInstances;
+        }
+
+        // get locality info of self-service and route by locality
+        Optional<XdsLocality> localityInfoOfSelfService = XdsRouterUtil.getLocalityInfoOfSelfService();
+        if (localityInfoOfSelfService.isPresent()) {
+            Set<ServiceInstance> serviceInstances = getServiceInstanceOfLocalityCluster(clusterLoadAssigment,
+                    localityInfoOfSelfService.get());
+            if (!serviceInstances.isEmpty()) {
+                return serviceInstances;
+            }
+        }
+
+        Set<ServiceInstance> serviceInstances = getServiceInstanceOfCluster(clusterLoadAssigment);
+        return serviceInstances.isEmpty() ? serviceDiscovery.getServiceInstance(serviceName) : serviceInstances;
+    }
+
+    private Set<ServiceInstance> getServiceInstanceOfLocalityCluster(XdsClusterLoadAssigment clusterLoadAssigment,
+            XdsLocality locality) {
+        return clusterLoadAssigment.getLocalityInstances().entrySet().stream()
+                .filter(xdsLocalitySetEntry -> xdsLocalitySetEntry.getKey().equals(locality))
+                .flatMap(xdsLocalitySetEntry -> xdsLocalitySetEntry.getValue().stream())
+                .collect(Collectors.toSet());
+    }
+
+    private Set<ServiceInstance> getServiceInstanceOfCluster(XdsClusterLoadAssigment clusterLoadAssigment) {
+        return clusterLoadAssigment.getLocalityInstances().entrySet().stream()
+                .flatMap(instanceEntry -> instanceEntry.getValue().stream())
+                .collect(Collectors.toSet());
+    }
+
+    private boolean isPathMatched(XdsPathMatcher matcher, String path) {
+        return matcher.isMatch(path);
+    }
+
+    private boolean isHeadersMatched(List<XdsHeaderMatcher> matchers, Map<String, String> headers) {
+        return matchers.stream()
+                .allMatch(xdsHeaderMatcher -> xdsHeaderMatcher.isMatch(headers));
+    }
+
+    private String selectClusterByWeight(XdsWeightedClusters weightedClusters) {
+        List<XdsClusterWeight> clusters = weightedClusters.getClusters();
+        int totalWeight = weightedClusters.getTotalWeight();
+        if (CollectionUtils.isEmpty(clusters) || totalWeight == 0) {
+            return StringUtils.EMPTY;
+        }
+        int randomWeight = random.nextInt(totalWeight);
+
+        int currentWeight = 0;
+        for (XdsClusterWeight clusterWeight : clusters) {
+            currentWeight += clusterWeight.getWeight();
+            if (randomWeight < currentWeight) {
+                return clusterWeight.getClusterName();
+            }
+        }
+        return StringUtils.EMPTY;
+    }
+
+    /**
+     * XdsRouterHandler
+     *
+     * @author daizhenyu
+     * @since 2024-08-29
+     **/
+    public enum MatchType {
+        /**
```

Comment:
if you want to use in this class only，use private. Or move it outside

Reply:
done